src/utils/lockfree.hpp

```diff
@@ -0,0 +1,121 @@
+/**
+ * Canary - A free and open-source MMORPG server emulator
+ * Copyright (©) 2019-2024 OpenTibiaBR <opentibiabr@outlook.com>
+ * Repository: https://github.com/opentibiabr/canary
+ * License: https://github.com/opentibiabr/canary/blob/main/LICENSE
+ * Contributors: https://github.com/opentibiabr/canary/graphs/contributors
+ * Website: https://docs.opentibiabr.com/
+ */
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <vector>
+#include <thread>
+#include <spdlog/spdlog.h>
+
+const int TOTAL_THREADS = static_cast<int>(std::thread::hardware_concurrency());
+const size_t LOCAL_CACHE_LIMIT = std::max(35 / TOTAL_THREADS, 5);
+constexpr size_t STATIC_PREALLOCATION_SIZE = 100;
+
+struct StackNode {
+	void* data;
+	StackNode* next;
+};
+
+template <std::size_t TSize, size_t CAPACITY>
+class LockfreeFreeList {
+public:
+	LockfreeFreeList() :
+		head(nullptr) { }
+
+	static LockfreeFreeList &get() {
+		static LockfreeFreeList instance;
+		return instance;
+	}
+
+	bool pop(void*&result) {
+		StackNode* old_head = head.load(std::memory_order_acquire);
+		while (old_head != nullptr) {
+			if (head.compare_exchange_weak(old_head, old_head->next, std::memory_order_acquire)) {
+				result = old_head->data;
+				delete old_head;
+				return true;
+			}
+
+			std::this_thread::yield();
+		}
+		return false;
+	}
+
+	bool push(void* data) {
+		auto* new_node = new StackNode { data, nullptr };
+		StackNode* old_head = head.load(std::memory_order_relaxed);
+		do {
+			new_node->next = old_head;
+		} while (!head.compare_exchange_weak(old_head, new_node, std::memory_order_release, std::memory_order_relaxed));
```

Comment:
How do you plan on tackling the ABA problem here and in pop?

https://en.wikipedia.org/wiki/ABA_problem